README.md

```diff
@@ -242,6 +242,33 @@ function DelayedEffect(props: { timerMs: number }) {
 }
 ```
 
+**useReducer**
+
+You can use [Discriminated Unions](https://www.typescriptlang.org/docs/handbook/advanced-types.html) for reducer actions. Don't forget to define the return type of reducer, otherwise Typescript will infer it.
```

Comment:
This probably needs some clarification. One of the strengths of typescript is type inference. Advocating against it without additional explanation might be confusing.

I guess this is important to prevent accidental type widening? [The example used here](https://www.typescriptlang.org/play/index.html#src=type%20Action%20%3D%0D%0A%20%20%7B%20type%3A%20'SET_ONE'%3B%20payload%3A%20string%3B%20%7D%0D%0A%20%20%7C%20%7B%20type%3A%20'SET_TWO'%3B%20payload%3A%20number%3B%20%7D%3B%0D%0A%0D%0Ainterface%20AppState%20%7B%0D%0A%20%20one%3A%20string%3B%0D%0A%20%20two%3A%20number%3B%0D%0A%7D%0D%0A%0D%0Aexport%20function%20reducer(state%3A%20AppState%2C%20action%3A%20Action)%20%7B%0D%0A%20%20switch%20(action.type)%20%7B%0D%0A%20%20%20%20case%20'SET_ONE'%3A%0D%0A%20%20%20%20%20%20return%20%7B%0D%0A%20%20%20%20%20%20%20%20...state%2C%0D%0A%20%20%20%20%20%20%20%20one%3A%20action.payload%20%2F%2F%20%60payload%60%20is%20string%0D%0A%20%20%20%20%20%20%7D%3B%0D%0A%20%20%20%20case%20'SET_TWO'%3A%0D%0A%20%20%20%20%20%20return%20%7B%0D%0A%20%20%20%20%20%20%20%20...state%2C%0D%0A%20%20%20%20%20%20%20%20two%3A%20action.payload%20%2F%2F%20%60payload%60%20is%20number%0D%0A%20%20%20%20%20%20%7D%3B%0D%0A%20%20%20%20default%3A%0D%0A%20%20%20%20%20%20return%20state%3B%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20newState%20%3D%20reducer(%7Bone%3A%20'1'%2C%20two%3A%202%7D%2C%20%7Btype%3A%20'SET_ONE'%2C%20payload%3A%20'1'%7D)) is working fine with regards to type inference: hovering over `newState` shows `AppState` as the inferred type.

Reply:
What I mean is [this](https://www.typescriptlang.org/play/index.html#src=type%20Action%20%3D%0D%0A%20%20%7B%20type%3A%20'SET_ONE'%3B%20payload%3A%20string%3B%20%7D%0D%0A%20%20%7C%20%7B%20type%3A%20'SET_TWO'%3B%20payload%3A%20number%3B%20%7D%3B%0D%0A%0D%0Ainterface%20AppState%20%7B%0D%0A%20%20one%3A%20string%3B%0D%0A%20%20two%3A%20number%3B%0D%0A%7D%0D%0A%0D%0Aexport%20function%20reducer(state%3A%20AppState%2C%20action%3A%20Action)%20%7B%0D%0A%20%20switch%20(action.type)%20%7B%0D%0A%20%20%20%20case%20'SET_ONE'%3A%0D%0A%20%20%20%20%20%20return%20%7B%0D%0A%20%20%20%20%20%20%20%20...state%2C%0D%0A%20%20%20%20%20%20%20%20one%3A%20123%2C%0D%0A%20%20%20%20%20%20%20%20oneTest%3A%20action.payload%20%2F%2F%20%60payload%60%20is%20string%0D%0A%20%20%20%20%20%20%7D%3B%0D%0A%20%20%20%20case%20'SET_TWO'%3A%0D%0A%20%20%20%20%20%20return%20%7B%0D%0A%20%20%20%20%20%20%20%20...state%2C%0D%0A%20%20%20%20%20%20%20%20twoTest%3A%20action.payload%20%2F%2F%20%60payload%60%20is%20number%0D%0A%20%20%20%20%20%20%7D%3B%0D%0A%20%20%20%20default%3A%0D%0A%20%20%20%20%20%20return%20state%3B%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20newState%20%3D%20reducer(%7Bone%3A%20'1'%2C%20two%3A%202%7D%2C%20%7Btype%3A%20'SET_ONE'%2C%20payload%3A%20'1'%7D)).

However I am not sure what is the best way to explain this.

Reply:
This applies to any function signature though. As soon as you would assign `newState` to a value that expects `AppState` you get an error. 

I'm not disagreeing with your point. It's just not specific to reducers as far as I can tell.

Good compromise would be explain why the explicit annotation might be a good idea here. Maybe there's already a good resource that explains why explicit returntype annotation is a good idea.

Reply:
What about using something like this?

"Explicit types for function return values makes it clear to any calling code what type is returned. This ensures that the return value is assigned to a variable of the correct type."

Reply:
I'm not entirely convinced. I would really appreciate a comprehensive text about pros and cons of explicit return types. However, it's not obvious to me that this is the right place for such a text. Let's hear what other contributors have to say about this.

Reply:
It's not really our place to enforce TypeScript best practices IMO (though @sw-yx may think otherwise), but rather React + TypeScript best practices. In addition there's already an existing lint rule that's part of the recommended set of options in TSLint: https://github.com/bradzacher/eslint-plugin-typescript/blob/master/docs/rules/explicit-function-return-type.md

Reply:
i have generally operated without them, but i understand the desire for them and do use them myself only when i feel it helps. most of the examples here do not have explicit return types, though we do comment the types of results. i feel this is a nice balance of verbose vs explicit.

since this is a general TS practice my inclination is that discussion of this belongs in its own section, and not in useReducer at all.

Reply:
So what is the conclusion? I could simply remove the "Don't forget to define the return type.." sentence.

Reply:
Sounds good. Guarding against excess props is a more general concern and from my experience people can get very upset if you suggest that this helps catching bugs because it implies they write buggy code.

Reply:
haha. i think we have to make clear that we're just trying to get people started here, people are free to deviate from these recommendations for a number of equally valid reasons.